### PR TITLE
Small change to QA email view

### PR DESCRIPTION
**Pull request by raamSoftwire**
## Context

As a content designer, I can't see the reference name of the email so I have to jump back and forth. Also, without the name, if you're looking at a number of different emails it gets confusing.

## Changes proposed in this pull request

Add the mailer and email reference names to the email preview header

Before

image

After

image

## Guidance to review

- commit by commit 
- the first commit is a straight copy from the Rails repo
- do we need tests here?
- this change isn't scoped by environment. Is that a problem?

## Link to Trello card

https://trello.com/c/5Lf2AxrW/3033-not-a-small-change-to-qa-email-view

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

**Comment by raamSoftwire** [comment]
@malcolmbaig the `render_preview_wrapper` method in the `mail-notify` gem is `private`. Should we be able to do this? Does this surprise you?

**Comment by Tomcowan1** [comment]
Looks good to me, thanks Raam!

**Review by Tomcowan1 — APPROVED** [review]

File: spec/mailers/previews/email.html.erb [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+      width: 120px;
+      padding: 1px;
+      float: left;
+      clear: left;
+      text-align: right;
+      color: #7f7f7f;
+    }
+
+    dd {
+      margin-left: 90px; /* 80px + 10px */
```
Comment (paulrobertlloyd) [inline spec/mailers/previews/email.html.erb:40]:
```suggestion
      margin-left: 125px; /* 120px + 5px */
```

File: spec/mailers/previews/email.html.erb [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    dl {
+      margin: 0 0 10px 0;
+      padding: 0;
+    }
+
+    dt {
+      width: 120px;
+      padding: 1px;
```
Comment (paulrobertlloyd) [inline spec/mailers/previews/email.html.erb:32]: Let’s increase the padding a tiny bit 🤏

```suggestion
      padding: 2px;
```

File: spec/mailers/previews/email.html.erb [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      float: left;
+      clear: left;
+      text-align: right;
+      color: #7f7f7f;
+    }
+
+    dd {
+      margin-left: 90px; /* 80px + 10px */
+      padding: 1px;
```
Comment (paulrobertlloyd) [inline spec/mailers/previews/email.html.erb:41]: Let’s increase the padding a tiny bit 🤏

```suggestion
      padding: 2px;
```

File: spec/mailers/previews/email.html.erb [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    <% if @email.respond_to?(:smtp_envelope_to) && @email.to != @email.smtp_envelope_to %>
+      <dt>SMTP-To:</dt>
+      <dd><%= @email.smtp_envelope_to %></dd>
+    <% end %>
+
+    <dt>Mailer:</dt>
+    <dd><%= @email.instance_variable_get(:@mailer_class).instance_variable_get(:@mailer_name).humanize %></dd>
+
+    <dt>Reference name:</dt>
+    <dd><%= @email.instance_variable_get(:@action) %></dd>
```
Comment (paulrobertlloyd) [inline spec/mailers/previews/email.html.erb:72]: Can we combine these two fields to reduce the overall height of this header?

```suggestion
    <dt>Template:</dt>
    <dd><%= @email.instance_variable_get(:@action) %> (<%= @email.instance_variable_get(:@mailer_class).instance_variable_get(:@mailer_name).humanize %>)</dd>
```

File: spec/mailers/previews/email.html.erb (hunk 1 above)
Comment (paulrobertlloyd) [reply spec/mailers/previews/email.html.erb:40]: Or leave as is if you decide to accept the other recommendation which combines the two new fields under a single ‘Template’ key

File: spec/mailers/previews/email.html.erb [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+    <dt>From:</dt>
+    <dd><%= @email.header['from'] %></dd>
```
Comment (paulrobertlloyd) [inline spec/mailers/previews/email.html.erb:75]: Can re remove this given it’s always empty?

File: spec/mailers/previews/email.html.erb [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    <% if I18n.available_locales.count > 1 %>
+      <dt>Locale:</dt>
+      <dd>
+        <select id="locale" onchange="refreshBody(true);">
+          <% I18n.available_locales.each do |locale| %>
+            <option <%= I18n.locale == locale ? 'selected' : '' %> value="<%= locale_query(locale) %>"><%= locale %></option>
+          <% end %>
+        </select>
+      </dd>
+    <% end %>
```
Comment (paulrobertlloyd) [inline spec/mailers/previews/email.html.erb:127]: Do we need to show these values?

**Comment by paulrobertlloyd** [comment]
Suggested a few tweaks, after which the header will look more like this:

Screenshot 2021-02-23 at 18 32 17

File: spec/mailers/previews/email.html.erb (hunk 1 above)
Comment (raamSoftwire) [reply spec/mailers/previews/email.html.erb:40]: I've accepted the other change and left this one.

**Comment by raamSoftwire** [comment]
@paulrobertlloyd Cheers for the styling updates, it looks much neater now

File: spec/mailers/previews/email.html.erb (hunk 6 above)
Comment (raamSoftwire) [reply spec/mailers/previews/email.html.erb]: Nope. I've removed them too

**Comment by malcolmbaig** [comment]
> @malcolmbaig the `render_preview_wrapper` method in the `mail-notify` gem is `private`. Should we be able to do this? Does this surprise you?

Ideally the gem would expose some public mechanism for us to declare which template is used for the preview wrapper. In the absence of that, I think overriding the private method is fine - especially as the worst that can happen is that previews start breaking after a mail-notify version update. 

An alternative option with a smaller footprint—and worth considering applying here—is the following:

- Add `prepend_view_path('spec/mailers/previews')` just before this line https://github.com/DFE-Digital/apply-for-teacher-training/commit/e901fbad662db5b08faa18652b03e582c4b7e5f4#diff-e530c57fe094b50f394218dfd0b938dda863d461c8a1d61301dece5925367048R4
- Rename our custom template to `spec/mailers/previews/rails/mailers/email.html.erb`
- Remove the `render_preview_wrapper` override.

The `render_preview_wrapper` in `mail-notify` will then look for the email template in our directory before searching the default view paths.

File: spec/mailers/previews/email.html.erb [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+      overflow: hidden;
+    }
+
+    dl {
+      margin: 0 0 10px 0;
+      padding: 0;
+    }
+
+    dt {
+      width: 120px;
```
Comment (paulrobertlloyd) [inline spec/mailers/previews/email.html.erb:31]: Just need to revert this to the width used by the original Rails component.

```suggestion
      width: 80px;
```

Comment (raamSoftwire) [reply spec/mailers/previews/email.html.erb]: Done

**Comment by raamSoftwire** [comment]
@malcolmbaig thanks again for the help with this. I've made those changes now

**Review by paulrobertlloyd — APPROVED** [review]

**Review by malcolmbaig — APPROVED** [review]

**Pull request merged**